Connection Lifecycle Management Infrastructure

anarazel opened the pull request:
Subset of #775, for easier review.

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,541 @@
+ * This is called by Citus' global transaction callback.
+ */
+void
+AtEOXact_Connections(bool isCommit)
```
Maybe PurgeTemporaryConnections?

anarazel replied:
Well, it does more than that, it also check for session connections and verifies that they're in an ok state and closes them otherwise.

marcocitus replied:
ResetConnections?

anarazel replied:
I settled, for now, on AfterXactConnectionHandling && AfterXactHostConnectionHandling (for the out of line portion you wanted).

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,541 @@
+		 * Have to iterate "manually", to be able to delete connections in the
+		 * middle of the list.
+		 */
+		for (connectionCell = list_head(entry->connections);
```
For readability, could we move the inner-loop into its own function?

anarazel replied:
I did, but it doesn't really seem like a beneficial change, especially due to the "forced" placement of static functions... :(

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,541 @@
+ * If user or database are NULL, the current session's defaults are used. The
+ * following flags influence connection establishment behaviour:
+ * - NEW_CONNECTION - it is permitted to establish a new connection
+ * - CACHED_CONNECTION - it is permitted to re-use an established connection
```
For simplicity, I'd prefer to remove these flags until there is a concrete need for them. As a caller of this API, I don't care that much whether my connection is cached or new. Even in the overall PR, all the callers set both flags right now.

In general, it seems like callers will mainly care a lot about what state the connection is in (e.g. I might want a connection that is not claimed, not part of coordinated transaction, has no custom SETs), so those are probably appropriate flags at some point.

anarazel replied:
Hm, ok.

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,541 @@
+				 * FIXME: This should be removed soon, once all connections go
+				 * through this API.
+				 */
+				if (!connection->activeInTransaction &&
```
Maybe omit the activeInTransaction flag, it doesn't seem to be used (always true here?) and not quite clear how we could know that here.

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,541 @@
+
+	if (flags & SESSION_LIFESPAN)
+	{
+		connection->sessionLifespan = true;
```
For simplicity, requiring to caller to think about whether a connection should have a session lifespan seems undesirable. It currently happens to work because the router executor doesn't claim connections exclusively, uses appropriate caching flags, and therefore under normal conditions either always reuses a connection or opens a new one using session lifespan, and there are no other callers that use the flag, meaning in the end we cache 1 session lifespan connection. That's a lot of stuff to have to understand to know how we cache connections and likely to be broken at some point if we force that complexity onto callers.

It'd be nicer if we hide that complexity by implementing a simple caching policy, e.g.: "the first connection to a node always gets sessionLifespan" and revising it when the need arises.

anarazel replied:
I very much disagree. It's completely caller specific whether connections should be session lifespan. DDL should pretty much never (DROP TABLE;), others sometimes (COPY when flag specified), others always. There's no way to magically specify this.

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,541 @@
+/*
+ * Synchronously finish connection establishment of an individual connection.
+ *
+ * TODO: Replace with variant waiting for multiple connections.
```
Can't wait for the TODO ;)

anarazel replied:
Easy enough to implement, would you like this to be tackled in this PR?

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,541 @@
+ * TODO: Replace with variant waiting for multiple connections.
+ */
+void
+FinishConnectionEstablishment(MultiConnection *connection)
```
I know we have some wonky calling order rules, but maybe break them here such that this can be just below StartConnectionEstablishment.

anarazel replied:
I'm all for that. I find the current rules pretty much an anti-pattern.

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,541 @@
+			}
+			else if (pollResult > 0)
+			{
+				/* IO possible, continue connection establishment */
```
s/IO/if/

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,541 @@
+			}
+			else if (pollResult != EINTR)
+			{
+				/* retrying, signal */
```
Should we check for cancellation at some point in this while loop? (not necessarily here)

anarazel replied:
The problem is that doing so reliably and portably essentially requires doing something like postgres' latches do, since poll/select aren't guaranteed to be interrupted by signals. In 9.6 this is trivially implementable for multiple connections because we/I introduced WaitEventSets which can wait both on latches and multiple ports at the same time (https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/include/storage/latch.h;hb=HEAD).

If we just use a single connection, we can just use WaitLatchOrSocket, but if we want to wait on multiple ones, we have to wait for 9.6 :(

marcocitus commented on src/include/distributed/connection_management.h:
```diff
@@ -0,0 +1,133 @@
+	bool sessionLifespan;
+
+	/* is the connection currently in use, and shouldn't be used by anything else */
+	bool claimedExclusively;
```
This name makes sense to me.

marcocitus commented on src/include/distributed/connection_management.h:
```diff
@@ -0,0 +1,133 @@
+	int32 port;
+	char user[NAMEDATALEN];
+	char database[NAMEDATALEN];
+} ConnectionHashKey;
```
Unimportant: I was thinking that we could maybe give this a generic name (NodeURL?)

anarazel replied:
Hm, most other places probably shouldn't store user/database "inline". What could you see this being used for otherwise?

marcocitus commented on src/include/distributed/connection_cache.h:
```diff
@@ -52,26 +52,11 @@ typedef struct NodeConnectionEntry
 extern PGconn * PurgeConnectionByKey(NodeConnectionKey *nodeConnectionKey);
-extern bool SqlStateMatchesCategory(char *sqlStateString, int category);
 extern void WarnRemoteError(PGconn *connection, PGresult *result);
 extern void ReraiseRemoteError(PGconn *connection, PGresult *result);
```
Should maybe move into a remote_commands.h file.

anarazel replied:
Which of these? You mean WarnRemoteError/ReraiseRemoteError? There's a variant of them in remote_commands, but it's a slightly different API. Partially because when I wrote that code we still purged connections upon errors...

anarazel replied:
Right.

anarazel commented:
> Two general questions: What happens if the router executor now calls GetNodeConnection instead of GetOrEstablishConnection?

There's no integration with the router executor (or any other user) as of this moment. If it just plainly switched around, things should mostly work I think, except that we'd need to change/remove the purging of connections and such.

>  How do we close a connection before the end of the transaction?

At the moment, we don't. I had a function for it and removed it again, because I couldn't find a use-case.

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,541 @@
+			}
+			pollFileDescriptor.revents = 0;
+
+			pollResult = poll(&pollFileDescriptor, 1, CLIENT_CONNECT_TIMEOUT_SECONDS_INT);
```
The timeout argument should be specified in milliseconds, not seconds.

marcocitus commented:
> At the moment, we don't. I had a function for it and removed it again, because I couldn't find a use-case.

We should typically close individual connections after a statement that uses the real-time executor. We currently also close connections before the end of the transaction when a non-critical network failure occurs, which makes sense because you don't want to keep ports in use for no reason. Using a separate, global hash for all connections would make it easier to remove connections.

marcocitus commented:
> There's no integration with the router executor (or any other user) as of this moment. If it just plainly switched around, things should mostly work I think, except that we'd need to change/remove the purging of connections and such.

I think it would be useful if we convert the router executor such that we can remove the connection cache code that these functions replace and test them.

anarazel commented:
> >  At the moment, we don't. I had a function for it and removed it again, because I couldn't find a use-case.

> We should typically close individual connections after a statement that uses the real-time executor. We currently also close connections before the end of the transaction when a non-critical network failure occurs, which makes sense because you don't want to keep ports in use for no reason.

I can add one, but I'm unconvinced it really helps. Not caring and simply letting end-of-transaction taking care of it seems a lot simpler.

> Using a separate, global hash for all connections would make it easier to remove connections.

There's no technical difficulty in doing so, I just didn't see a use-case.

anarazel commented:
> I think it would be useful if we convert the router executor such that we can remove the connection cache code that these functions replace and test them.

Even a minimal conversion of router is fairly large, as it now has interactions with multi_shard_transaction.[ch], due to ExecuteModifyTasks().  How about one of the very simple conversions, say WorkerShardStat() or such instead?

anarazel commented:
@marcocitus I've pushed a revised version. To be called out:
- Replaced pg_list.h singly linked list with doubly linked list, looks a lot better imo
- Added a minimal CloseConnection()
-  re-added the CACHED_CONNECTION (and thus also the NEW_CONNECTION) flags. They were necessary for:
- As a demonstrator / validator I've made connection_cache.[ch]'s GetOrEstablishConnection/PurgeConnection/PurgeConnectionByKey/ be a layer over connection_management.c.  We can't unfortunately do the same for ConnectToNode() as its callers close connections themselves.
Previously I'd pushed:
- changed timeout implementation so it warns after a timeout, and made it check interrupts in a regular interval
- smaller stuff

marcocitus commented on src/include/distributed/commit_protocol.h:
```diff
@@ -14,18 +14,12 @@
-{
-	COMMIT_PROTOCOL_1PC = 0,
-	COMMIT_PROTOCOL_2PC = 1
-} CommitProtocolType;
```
Any reason for moving this from commit_protocol.h?

anarazel replied:
We can delay it, but for me it seems a lot more appropriate to have that in transaction_management.h. You want to push that back?

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,612 @@
+ * This is called by Citus' global transaction callback.
+ */
+void
+AfterXactConnectionHandling(bool isCommit)
```
Not sure what this does from the name or the description. Maybe AfterXactResetConnections?

anarazel commented:
Hm. I think an even better idea would be to add a transitional connection mgmt.h API for closing connections via a pgconn.

That'd allow us to move all connection handling below the new framework, with nothing but a few s/PQfinish/CloseConnectionUsingPgconn/.


-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

marcocitus commented on src/backend/distributed/utils/connection_cache.c:
```diff
@@ -60,56 +51,20 @@ static void ReportRemoteError(PGconn *connection, PGresult *result, bool raiseEr
-	bool entryFound = false;
-	bool needNewConnection = true;
-	char *userName = CurrentUserName();
+	MultiConnection *mconnection =
```
Probably needs some better naming, e.g. multiConnection. Maybe also conn -> pqConn

marcocitus commented:
> Hm. I think an even better idea would be to add a transitional connection mgmt.h API for closing connections via a pgconn.

Works for me.

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,612 @@
+
+/*
+ * UnclaimConnection signals that this connection is not being used
+ * anymore. That means it again may be returned by returned by
```
s/returned by returned by/returned by/

anarazel replied:
I think so, yes (done).

marcocitus commented on src/backend/distributed/utils/connection_cache.c:
```diff
@@ -60,56 +51,20 @@ static void ReportRemoteError(PGconn *connection, PGresult *result, bool raiseEr
-	bool needNewConnection = true;
-	char *userName = CurrentUserName();
+	MultiConnection *mconnection =
+		GetNodeConnection(connectionFlags, nodeName, nodePort);
```
Are we in a position to remove the call to PurgeConnection from ReportRemoteError now?

anarazel commented on src/backend/distributed/utils/connection_cache.c:
```diff
@@ -174,57 +130,24 @@ BuildKeyForConnection(PGconn *connection, NodeConnectionKey *connectionKey)
+	MultiConnection *connection;
+
+	connection =
+		StartNodeUserDatabaseConnection(
```
I've added a proper API (based on your implementation, but made a bit more general) now.

anarazel commented:
@marcocitus Updated:
- I've adapted a good bit of your changes
- I've amended (c.f. the to-be-squashed fixup! commits) the connection_cache conversion, so all its users now go through the new management; that required replacing a bunch of PQfinish()es in callers.
- I also pushed a new commit that bases multi_client_executor.[ch] onto of connection management. That actually works suprisingly well, and gets rid of a number of warts.  Still a pain to use, but ... ;)

This also shows why I want the flags...

What do you think?

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,687 @@
+		{
+			connection->sessionLifespan = true;
+		}
+		connection->activeInTransaction = true;
```
The 5 lines above don't belong in a "Find" function. Also, it's not very clear to me why we have this activeInTransaction flag or what it means to skip over it when `!connection->activeInTransaction && XactModificationLevel > XACT_MODIFICATION_DATA`.

anarazel replied:
Consider the scenario where you have connections open to some nodes, but you haven't used them in a transaction. If we're e.g. doing router executor type transactions, we're not allowed to open new connections after the beginning of a transactions. So we have to keep track whether connections already have been used in a transaction or not.

I agree that this would be better placed in the caller.

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,687 @@
+	else
+	{
+		/* XXX: we could error out instead */
+		ereport(WARNING, (errmsg("closing untracked connection")));
```
erroring out seems slight preferable to me (or maybe Assert(false)?)

anarazel replied:
WFM.

marcocitus commented:
Unsurprisingly, I still think the current flags are redundant and make the code harder to understand. In order to figure out how different connections interact and which flags to use, I have to figure out all the different ways in which StartNodeUserDatabaseConnection is called and how those calls combine. The way in which the flags are defined means the outcome of each call depends on the current state of the connection pool as well.

Perhaps part of the problem is that the current API is trying to do too many things. My main expectation for  connection lifecycle management is: Ensure that we close all open connections on abort, except keep one connection open between transactions.

I think there's some vaguely defined behaviour for caching connections within the same transaction, but that seems better handled at a higher layer. Same with the transaction behaviour.

Instead of having `CACHED_CONNECTION` and `NEW_CONNECTION` the API would be a lot simpler if we just had `FORCE_NEW_CONNECTION`:
https://github.com/citusdata/citus/commit/99fc0b06e75c69b93d09ec2d9c01660be880681f

(I'd actually have a slight preference to use the `CLAIM_EXCLUSIVELY` flag which sets `claimExclusively = true` instead of `FORCE_NEW_CONNECTION`, to further reduce ambiguity between those concepts.)

Additionally, we can rather easily remove the session lifespan flag by introducing some global policy:
https://github.com/citusdata/citus/commit/65267b429c4673daa4bd3d715267f939f0b9d1e3 . This creates a small behavioural change, but that seems perfectly acceptable. Compare it to pgbouncer requiring clients to specify the behaviour on every call. It might be more optimal in some cases, but no one would use pgbouncer.

These are small and trivial changes that drastically reduce the number of different ways in which we can call this API, so it seems like a no-brainer to me.

anarazel commented on src/backend/distributed/executor/multi_client_executor.c:
```diff
@@ -87,76 +89,34 @@ int32
-	char *effectiveDatabaseName = NULL;
-	char *effectiveUserName = NULL;
-
-	if (XactModificationLevel > XACT_MODIFICATION_NONE)
```
I'm not particularly happy with that piece of logic, so I'm happy to move it around. But it's not generally clear how you can deal with it all the callsites, if you consider connection reusage making things safe.

anarazel commented:
> Unsurprisingly, I still think the current flags are redundant and make the code harder to understand because they are leaky abstractions whose behaviour affects and depends on the state of the connection pool. In order to figure out how different connections interact and which flags to use, I have to figure out all the different ways in which StartNodeUserDatabaseConnection is called and how those calls combine. I wouldn't have to do that if there were just one or two ways.

> Instead of having CACHED_CONNECTION and NEW_CONNECTION the API would be a lot simpler if we just had FORCE_NEW_CONNECTION:
99fc0b0

I don't quite see how a flag for "only use already established connection" and "allow use of cached connection" are that confusing. But I can live with your proposal of FORCE_NEW_CONNECTION.   If we need it, we can add FORCE_CACHED without any trouble.

> Perhaps part of the problem is that the current API is trying to do too many things. My main expectation for connection lifecycle management is: Ensure that we close all open connections on abort/commit, except keep one connection open between transactions.

> Additionally, we can rather easily remove the session lifespan flag by introducing some global policy:
65267b4 . This creates a small behavioural change, but that seems perfectly acceptable. 
I think that's one thing where we disagree, and which I don't see us resolving. I think it's a fundamental error to restrict the caching to exactly one connection.

(I'd actually have a slight preference to use the CLAIM_EXCLUSIVELY flag which sets claimExclusively = true instead of FORCE_NEW_CONNECTION, to further reduce ambiguity between those concepts.)

> These are small and trivial changes that drastically reduce the number of different ways in which this API can be called, so it seems like a no-brainer to me.

Well, the one-connection only bit isn't trivial at all, to me it's quite fundamental. The rest I agree with.

anarazel commented:
> Instead of having CACHED_CONNECTION and NEW_CONNECTION the API would be a lot simpler if we just had FORCE_NEW_CONNECTION:
99fc0b0

I've cherry-picked this into the relevant commits.  I'm quite unconvinced that it's a good idea to repeat the logic in the callers, but whatever.

marcocitus commented on src/backend/distributed/utils/remote_commands.c:
```diff
@@ -0,0 +1,195 @@
+
+
+/*
+ * Log commands send to remote nodes if citus.log_remote_commands wants us to
```
Not terribly important, but function comments should use the "<function name> does X" format.

anarazel replied:
fixed.

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,635 @@
+ * UnclaimConnection().
+ */
+void
+ClaimConnectionExclusively(MultiConnection *connection)
```
Do we still expect to use this? Right now it's dead code.

anarazel replied:
Yes.

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,635 @@
+/*-------------------------------------------------------------------------
+ *
+ * connection_management.c
```
Minor: I think our utils directory is a little cramped, might be nice to put this into a connection directory.

anarazel replied:
Hm. That'd then only have two files? Ar we forseeing many more?

marcocitus replied:
I'm guessing we'll at least have functions for dealing with many connections. Could also be part of executor/ maybe.

marcocitus commented on src/backend/distributed/transaction/transaction_management.c:
```diff
@@ -0,0 +1,145 @@
+			/*
+			 * FIXME: do we want to support this? Or error out? Might be
+			 * annoying to error out as it could prevent experimentation. If
+			 * we error out, we should only do so if a coordinated transaction
```
Erroring out if there is a coordinated transaction would make sense (e.g. router executor errors out here if XactModificationLevel is not XACT_MODIFICATION_DATA).

anarazel replied:
@marcocitus If you're ok with that, I'll leave the FIXME in there, so we can get this show going? I'd be ok with either leaving it with the fixme or opening an issue?

marcocitus replied:
No, there are currently no conflicts between MX operations (which work mainly on distributed tables and metadata) and other Citus operations (which work mainly on shards), so it doesn't use the XactModificationLevel. It does already error out when a PREPARE TRANSACTION occurs after a metadata syncing operation.

anarazel replied:
(I presume you meant >= IDLE?)

marcocitus replied:
yeh, or != NONE

anarazel replied:
done.

marcocitus commented on src/backend/distributed/utils/remote_commands.c:
```diff
@@ -0,0 +1,195 @@
+ * IO.
+ */
+void
+ForgetResults(MultiConnection *connection)
```
Maybe omit for now? It seems like this function needs work.

anarazel replied:
It's pretty similar to what we already do in various places, just making the issues it has apparent :(. I'd rather centralize that.

marcocitus replied:
What does it replace? It's not called anywhere yet.

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,635 @@
+
+		/*
+		 * FIXME: Do we want to add transparent retry support here?
+		 */
```
I suppose we do have a minor regression compared to ConnectToNode in that regard. I think where it becomes relevant is when we're doing an operation like COPY and making 256 connections, but time out on the 238th one and roll back everything because of it. Would it be hard to add? Wouldn't mind doing it in a separate PR.

marcocitus commented on src/include/distributed/transaction_management.h:
```diff
@@ -0,0 +1,59 @@
+
+
+/*
+ * Initialization.
```
This comment is most enlightening.

anarazel replied:
Well, there'll later be more functions. And those that "belong" together have a comment grouping them together. I can remove it, but I see little reason to do so.

marcocitus replied:
Maybe `/* initialization functions */`

marcocitus commented on src/backend/distributed/utils/remote_commands.c:
```diff
@@ -0,0 +1,195 @@
+	{
+		PGresult *result = NULL;
+		result = PQgetResult(connection->conn);
+		if (result == NULL)
```
Maybe `while ((result = PQgetResult(connection->conn) != NULL)`

marcocitus commented on src/backend/distributed/utils/remote_commands.c:
```diff
@@ -0,0 +1,195 @@
+		{
+			PQputCopyEnd(connection->conn, NULL);
+
+			/* FIXME: mark connection as failed? */
```
Not quite sure what this means.

anarazel replied:
edited a teensy bit.

marcocitus commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,635 @@
+
+
+/*
+ * StartNodeConnection initiate a connection to remote node, using default
```
initiate->initiates

anarazel commented:
On 2016-12-06 13:15:51 -0800, Marco Slot wrote:
> +	/* underlying libpq connection */
> +	struct pg_conn *conn;
> 
> I'd have a slight preference for calling this something like `pqConn` or `pgConn` to make it easier to distinguish it from MultiConnection variables.

Heh. Then I can never remember which of those it is :P

anarazel commented on src/backend/distributed/utils/connection_management.c:
```diff
@@ -0,0 +1,635 @@
+ * If user or database are NULL, the current session's defaults are used. The
+ * following flags influence connection establishment behaviour:
+ * - FORCE_NEW_CONNECTION - a new connection is required
+ *
```
.oO(who removed that :P)

anarazel replied:
.oO(who removed that :P)

anarazel replied:
done

anarazel commented on src/include/distributed/connection_management.h:
```diff
@@ -0,0 +1,127 @@
+	char database[NAMEDATALEN];
+
+	/* underlying libpq connection */
+	struct pg_conn *conn;
```
done (pgConn)

marcocitus commented on src/backend/distributed/utils/connection_cache.c:
```diff
@@ -60,78 +51,27 @@ static void ReportRemoteError(PGconn *connection, PGresult *result, bool raiseEr
 PGconn *
 GetOrEstablishConnection(char *nodeName, int32 nodePort)
 {
+	int connectionFlags = SESSION_LIFESPAN;
```
I'm ok with moving ahead, but noting that I still find the use of this flag pretty weird and it will likely cause bugs further down the line. The API currently doesn't guarantee that there is only a single sessionLifespan connection. For example, you could open a regular connection, claim it exclusively, open a new `SESSION_LIFESPAN` connection, unclaim the first connection, ask for a `SESSION_LIFESPAN` connection, and the API returns the first, now unclaimed connection with sessionLifespan marked as true. Now there are 2 session lifespan connections, which probably wasn't intended.

sumedhpathak commented:
Yay!

On Wed, Dec 7, 2016 at 11:51 AM, Andres Freund <notifications@github.com>
wrote:

> Merged #863 <https://github.com/citusdata/citus/pull/863>.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/citusdata/citus/pull/863#event-885808555>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACeeC4hxVZIV1bXHXea6Kxifc8YljQxjks5rFw4ogaJpZM4KRkgQ>
> .
>



-- 

Sumedh Pathak
Co-founder and VP Engineering | Citus Data | (650) 422 9797
Elastically scalable PostgreSQL as a service.
<https://www.citusdata.com/blog/2016/10/03/designing-your-saas-database-for-high-scalability/>
Learn how CloudFlare handles billions of events in real-time.
<https://blog.cloudflare.com/scaling-out-postgresql-for-cloudflare-analytics-using-citusdb/>
